libtenzir/src/plugin.cpp

```diff
@@ -636,6 +639,84 @@ auto aspect_plugin::aspect_name() const -> std::string {
   return name();
 }
 
+// -- parser plugin ------------------------------------------------------------
+
+auto plugin_parser::parse_strings(std::shared_ptr<arrow::StringArray> input,
+                                  operator_control_plane& ctrl) const
+  -> std::vector<std::pair<type, std::shared_ptr<arrow::Array>>> {
+  // TODO: Collecting finished table slices here is very bad for performance.
+  // For example, we have to concatenate new table slices. But there are also
+  // many questions with regards to semantics. This should be either completely
+  // rewritten or replaced with a different mechanism after the revamp.
+  auto output = std::vector<table_slice>{};
+  auto append_null = [&] {
+    if (output.empty()) {
+      auto schema = type{"tenzir.unknown", record_type{}};
+      output.emplace_back(arrow::RecordBatch::Make(schema.to_arrow_schema(), 1,
+                                                   arrow::ArrayVector{}),
+                          schema);
+      return;
+    }
+    auto& last = output.back();
+    auto null_builder = caf::get<record_type>(last.schema())
+                          .make_arrow_builder(arrow::default_memory_pool());
+    TENZIR_ASSERT_CHEAP(null_builder->AppendNull().ok());
+    auto null_array = std::shared_ptr<arrow::StructArray>{};
+    TENZIR_ASSERT_CHEAP(null_builder->Finish(&null_array).ok());
+    auto null_batch = arrow::RecordBatch::Make(
+      last.schema().to_arrow_schema(), 1, null_array->Flatten().ValueOrDie());
+    last
+      = concatenate({std::move(last), table_slice{null_batch, last.schema()}});
+  };
+  for (auto str : values(string_type{}, *input)) {
+    if (not str) {
+      append_null();
+      continue;
+    }
+    auto bytes = std::span<const std::byte>{
+      reinterpret_cast<const std::byte*>(str->data()), str->size()};
+    auto chunk = chunk::make(bytes, []() noexcept {});
+    auto instance = instantiate(
+      [](chunk_ptr chunk) -> generator<chunk_ptr> {
+        co_yield std::move(chunk);
+      }(std::move(chunk)),
+      ctrl);
+    if (not instance) {
+      append_null();
+      continue;
+    }
+    auto slices = collect(std::move(*instance));
+    slices.erase(std::ranges::remove_if(slices,
+                                        [](table_slice& x) {
+                                          return x.rows() == 0;
+                                        })
+                   .begin(),
+                 slices.end());
```

Comment:
`std::erase_if(slices, [](const table_slice& x) { return x.rows() == 0; });`

Free function `std::erase(_if)` does the remove+erase dance automatically (C++20)

Reply:
Ah, yes, I remember now. Thanks :pray: